### ISSUE TITLE
Improve timer precision on Linux

### ISSUE THREAD

**Issue by stjepang** [stjepang/smol#81]
Unfortunately, epoll only supports blocking with timeouts of millisecond precision (because `epoll_wait()` accepts milliseconds as an argument), which means our timers will therefore also have millisecond precision.

The solution to improving precision is: create a timerfd that fires whenever the timeout is, thus interrupting `epoll_wait()`.

**Comment by jjl** [comment]
This might be solved by moving to io_uring on linux, the benchmarks for which seem to be compelling.

**Comment by YorickPeterse** [comment]
@jjl The problem with io_uring is that it's only available on very recent kernel versions, and it will likely take 2-3 years before it's more widely available. This means that for Linux there probably have to be two backends: one using epoll/timerfd, and one using io_uring.

**Comment by seirdb** [comment]
Would this replace the current implementation, or sit behind a feature flag? Furthermore, would an additional dependency on `timerfd` be allowed?

**Comment by stjepang** [comment]
We don't need the `timerfd` dependency - in fact, I'm pretty sure using timerfd from the `nix` crate will be even easier :)

It doesn't need to sit behind a feature flag as it doesn't affect the public API.

**Comment by seirdb** [comment]
Regarding your _comment https://github.com/stjepang/smol/issues/113#issuecomment-629767218_ in an other issue; is this rework still necessary?

**Comment by stjepang** [comment]
Yeah, this is still necessary if we want finer-grained timers (<1 ms)
